src/core/qgsaction.cpp

```diff
@@ -74,6 +87,183 @@ void QgsAction::run( const QgsExpressionContext &expressionContext ) const
     else
       QDesktopServices::openUrl( QUrl( expandedAction, QUrl::TolerantMode ) );
   }
+  else if ( mType == QgsAction::SubmitUrlEncoded || mType == QgsAction::SubmitUrlMultipart )
+  {
+
+    QUrl url{ expandedAction };
+
+    // Encode '+' (fully encoded doesn't encode it)
+    const QString payload { url.query( QUrl::ComponentFormattingOption::FullyEncoded ).replace( QChar( '+' ), QStringLiteral( "%2B" ) ) };
+
+    // Remove query string from URL
+    const QUrlQuery queryString { url.query( ) };
+    url.setQuery( QString( ) );
+
+    QNetworkRequest req { url };
+
+    // Specific code for testing, produces an invalid POST but we can still listen to
+    // signals and examine the request
+    if ( url.toString().contains( QLatin1String( "fake_qgis_http_endpoint" ) ) )
+    {
+      req.setUrl( QStringLiteral( "file://%1" ).arg( url.path() ) );
+    }
+
+    QNetworkReply *reply = nullptr;
+
+    if ( mType != QgsAction::SubmitUrlMultipart )
+    {
+      QString contentType { QStringLiteral( "application/x-www-form-urlencoded" ) };
+      // check for json
+      QJsonParseError jsonError;
+      QJsonDocument::fromJson( payload.toUtf8(), &jsonError );
+      if ( jsonError.error == QJsonParseError::ParseError::NoError )
+      {
+        contentType = QStringLiteral( "application/json" );
+      }
+      req.setHeader( QNetworkRequest::KnownHeaders::ContentTypeHeader, contentType );
+      reply = QgsNetworkAccessManager::instance()->post( req, payload.toUtf8() );
+    }
+    // for multipart create parts and headers
+    else
+    {
+      QHttpMultiPart *multiPart = new QHttpMultiPart( QHttpMultiPart::FormDataType );
+      const auto queryItems { queryString.queryItems( QUrl::ComponentFormattingOption::FullyDecoded ) };
+      for ( const auto &queryItem : std::as_const( queryItems ) )
+      {
+        QHttpPart part;
+        part.setHeader( QNetworkRequest::ContentDispositionHeader,
+                        QStringLiteral( "form-data; name=\"%1\"" )
+                        .arg( QString( queryItem.first ).replace( '"', QStringLiteral( R"(\")" ) ) ) );
+        part.setBody( queryItem.second.toUtf8() );
+        multiPart->append( part );
+      }
+      reply = QgsNetworkAccessManager::instance()->post( req, multiPart );
+      multiPart->setParent( reply );
+    }
+
+    QObject::connect( reply, &QNetworkReply::finished, reply, [ reply ]
+    {
+      if ( reply->error() == QNetworkReply::NoError )
+      {
+
+        if ( reply->attribute( QNetworkRequest::RedirectionTargetAttribute ).isNull() )
+        {
+
+          const QByteArray replyData = reply->readAll();
+
+          QString filename { "download.bin" };
+          if ( const auto header = reply->header( QNetworkRequest::KnownHeaders::ContentDispositionHeader ).toString().toStdString(); ! header.empty() )
```

Comment:
there's so much auto use here that I can't review this code without pulling it into an IDE to actually see what's happening :sob:

Reply:
auto removed